File: frontend/src/utils/profitability.ts
// Единая шкала рентабельности — одна на все экраны, где показывается этот
// показатель: реестр проектов, карточка проекта, шапка версии бюджета и
// результаты расчёта.
export interface ProfitabilityGrade {
  color: string;
  label: string;
}

const SCALE: { min: number; grade: ProfitabilityGrade }[] = [
  { min: 30, grade: { color: '#2F7D3A', label: 'Сверхприбыльный' } },
  { min: 20, grade: { color: '#4E9455', label: 'Высокорентабельный' } },
  { min: 5, grade: { color: '#B07A12', label: 'Среднерентабельный' } },
  { min: 1, grade: { color: '#A8621A', label: 'Низкорентабельный' } },
  { min: 0, grade: { color: '#8E4A2A', label: 'Порог рентабельности' } },
];

const LOSS: ProfitabilityGrade = { color: '#9C2B2B', label: 'Убыточный' };

export function profitabilityGrade(n: number): ProfitabilityGrade {
  // «> 30 %» — строго больше, остальные пороги включают нижнюю границу.
  if (n > 30) return SCALE[0].grade;
  return SCALE.slice(1).find(s => n >= s.min)?.grade ?? LOSS;
}

File: backend/internal/budgets/profitability.go
package budgets

// Шкала рентабельности — та же, что на экранах платформы
// (frontend/src/utils/profitability.ts). Границы заданы владельцем
// 2026-08-27; проверяются сверху вниз, нижняя граница включается: 20 % —
// уже «высокорентабельный», 5 % — «среднерентабельный».
//
// ДУБЛИРОВАНИЕ ОСОЗНАННОЕ. Экран берёт шкалу из TypeScript, книга — из
// Go, и разойтись они не должны: вердикт в выгрузке обязан совпадать с
// цветом в реестре проектов. Меняя пороги или цвета здесь, поправить и
// там — файлы ссылаются друг на друга этими комментариями.
//
// Цвета без решётки: excelize ждёт RGB шестью символами.

type profitabilityLevel struct {
	// Min — нижняя граница ступени, %.
	Min float64
	// Label — вердикт, который печатается в книге.
	Label string
	// Color — заливка строки-вывода, RGB без «#».
	Color string
}

// profitabilityScale — от самой доходной ступени к самой слабой.
var profitabilityScale = []profitabilityLevel{
	{Min: 30, Label: "Сверхприбыльный", Color: "2F7D3A"},
	{Min: 20, Label: "Высокорентабельный", Color: "4E9455"},
	{Min: 5, Label: "Среднерентабельный", Color: "B07A12"},
	{Min: 1, Label: "Низкорентабельный", Color: "A8621A"},
	{Min: 0, Label: "Порог рентабельности", Color: "8E4A2A"},
}

var profitabilityLoss = profitabilityLevel{Label: "Убыточный", Color: "9C2B2B"}

// profitabilityGrade — ступень шкалы для рентабельности n (в процентах).
func profitabilityGrade(n float64) profitabilityLevel {
	// «> 30 %» — строго больше, остальные пороги включают нижнюю границу.
	if n > 30 {
		return profitabilityScale[0]
	}
	for _, lvl := range profitabilityScale[1:] {
		if n >= lvl.Min {
			return lvl
		}
	}
	return profitabilityLoss
}
